### ISSUE TITLE
Process started from NX Console detecting wrong platform

### ISSUE THREAD

**Issue by joeldbirch** [nrwl/nx-console#1810]
## Current Behavior

Since today, clicking the "play" icon button in the NX Console next to project target using Vite shows this error:

<details>
<summary>Open to see error</summary>
Error: 
You installed esbuild for another platform than the one you're currently using.
This won't work because esbuild is written with native code and needs to
install a platform-specific binary executable.

Specifically the "@esbuild/darwin-arm64" package is present but this platform
needs the "@esbuild/darwin-x64" package instead. People often get into this
situation by installing esbuild with npm running inside of Rosetta 2 and then
trying to use it with node running outside of Rosetta 2, or vice versa (Rosetta
2 is Apple's on-the-fly x86_64-to-arm64 translation service).

If you are installing with npm, you can try ensuring that both npm and node are
not running under Rosetta 2 and then reinstalling esbuild. This likely involves
changing how you installed npm and/or node. For example, installing node with
the universal installer here should work: https://nodejs.org/en/download/. Or
you could consider using yarn instead of npm which has built-in support for
installing a package on multiple platforms simultaneously.

If you are installing with yarn, you can try listing both "arm64" and "x64"
in your ".yarnrc.yml" file using the "supportedArchitectures" feature:
https://yarnpkg.com/configuration/yarnrc/#supportedArchitectures
Keep in mind that this means multiple copies of esbuild will be present.

Another alternative is to use the "esbuild-wasm" package instead, which works
the same way on all platforms. But it comes with a heavy performance cost and
can sometimes be 10x slower than the "esbuild" package, so you may also not
want to do that.
```bash
    at generateBinPath (/Users/me/galaxy/node_modules/.pnpm/esbuild@0.17.5/node_modules/esbuild/lib/main.js:1873:17)
    at esbuildCommandAndArgs (/Users/me/galaxy/node_modules/.pnpm/esbuild@0.17.5/node_modules/esbuild/lib/main.js:1954:33)
    at ensureServiceIsRunning (/Users/me/galaxy/node_modules/.pnpm/esbuild@0.17.5/node_modules/esbuild/lib/main.js:2118:25)
    at build (/Users/me/galaxy/node_modules/.pnpm/esbuild@0.17.5/node_modules/esbuild/lib/main.js:2010:26)
    at bundleConfigFile (file:///Users/me/galaxy/node_modules/.pnpm/vite@4.3.4_@types+node@18.14.2_less@4.1.3_sass@1.62.1_stylus@0.55.0/node_modules/vite/dist/node/chunks/dep-f7d05e3f.js:64323:26)
    at loadConfigFromFile (file:///Users/me/galaxy/node_modules/.pnpm/vite@4.3.4_@types+node@18.14.2_less@4.1.3_sass@1.62.1_stylus@0.55.0/node_modules/vite/dist/node/chunks/dep-f7d05e3f.js:64299:31)
    at resolveConfig (file:///Users/me/galaxy/node_modules/.pnpm/vite@4.3.4_@types+node@18.14.2_less@4.1.3_sass@1.62.1_stylus@0.55.0/node_modules/vite/dist/node/chunks/dep-f7d05e3f.js:63908:34)
    at _createServer (file:///Users/me/galaxy/node_modules/.pnpm/vite@4.3.4_@types+node@18.14.2_less@4.1.3_sass@1.62.1_stylus@0.55.0/node_modules/vite/dist/node/chunks/dep-f7d05e3f.js:63192:26)
    at Module.createServer (file:///Users/me/galaxy/node_modules/.pnpm/vite@4.3.4_@types+node@18.14.2_less@4.1.3_sass@1.62.1_stylus@0.55.0/node_modules/vite/dist/node/chunks/dep-f7d05e3f.js:63189:12)
    at /Users/me/galaxy/node_modules/.pnpm/vite@4.3.4_@types+node@18.14.2_less@4.1.3_sass@1.62.1_stylus@0.55.0/node_modules/vite/index.cjs:23:55
```
</details>

Please note that VS Code updated today which was prior to me noticing this issue, so maybe this is contributing somehow.

Also note that running the equivalent NX target via the command line (be it from VS Code or iTerm) works flawlessly, so it seems to be related to the process that NX Console spins up.

While attempting to address the above esbuild error—which I had seen before and is usually fixed by reinstalling node_modules, but not this time—I ensured that everything about my platform was _not_ installed via Rosetta 2, as I have an M2 chip.

After reinstalling nvm, node, pnpm, VS Code and NX Console, and ensuring they are on arm64, I reran `pnpm i` on my workspace and got the same error. Then, I ran `nx migrate latest`, after which the error has changed to this:

```
 >  NX   Missing Platform Dependency

   The Nx CLI could not find or load the native binary for your supported platform (darwin-x64).
   This likely means that optional dependencies were not installed correctly, or your system is missing some system dependencies.
   For more information please see https://nx.dev/recipes/ci/troubleshoot-nx-install-issues
```
I did follow the troubleshooting tips in that link and ensured I could find `@nx/nx-darwin-arm64` in my node_modules, but this did not fix the issue.

## Expected Behavior

Clicking the "play" icon button in the NX Console next to project targets should work like it did yesterday. Ie. it should spin up a process and run the equivalent command.

## Steps to Reproduce

1. Match my environment.
2. Use a generator to create a React application that uses Vite.
3. Attempt to use NX Console to run its `serve` target.
4. See "Missing Platform Dependency" error.

I was easily able to reproduce the issue by cloning https://github.com/nrwl/nx-examples and attempting to run the apps/cart:serve target via the NX Console. The same error occurred with Yarn, which it is set up to run by default, and also after I changed it to use pnpm.

```
 *  Executing task: pnpm exec nx run cart:serve 
 
 >  NX   Missing Platform Dependency

   The Nx CLI could not find or load the native binary for your supported platform (darwin-x64).
   This likely means that optional dependencies were not installed correctly, or your system is missing some system dependencies.
   For more information please see https://nx.dev/recipes/ci/troubleshoot-nx-install-issues

 *  The terminal process "/bin/zsh '-l', '-c', 'pnpm exec nx run cart:serve'" terminated with exit code: 1. 
 *  Terminal will be reused by tasks, press any key to close it. 
```

### Environment

- Platform:
> Chip Apple M2 Pro
MacOS 13.2.1 (22D68)

- Nx Console version:
> v18.4.0

- VSCode version (I tried others also, such as the Apple Silicon version and Insiders, with the same result):

> Version: 1.80.0 (Universal)
Commit: 660393deaaa6d1996740ff4880f1bad43768c814
Date: 2023-07-04T13:39:33.766Z
Electron: 22.3.14
ElectronBuildId: 21893604
Chromium: 108.0.5359.215
Node.js: 16.17.1
V8: 10.8.168.25-electron.0
OS: Darwin arm64 22.3.0

- NX Report:

>    Node   : 18.15.0
   OS     : darwin-arm64
   pnpm   : 8.6.6
   nx (global)        : 16.3.2
   nx                 : 16.5.0
   @nx/js             : 16.5.0
   @nx/jest           : 16.5.0
   @nx/linter         : 16.5.0
   @nx/workspace      : 16.5.0
   @nx/cypress        : 16.5.0
   @nx/devkit         : 16.5.0
   @nx/esbuild        : 16.5.0
   @nx/eslint-plugin  : 16.5.0
   @nx/plugin         : 16.5.0
   @nx/react          : 16.5.0
   @nx/vite           : 16.5.0
   @nx/webpack        : 16.5.0
   nx-cloud           : 16.0.5
   typescript         : 5.1.6

**Comment by Cammisuli** [comment]
Oh thats not fun. I have a M1 chip myself, and things are running fine for me with the nx examples repo. 

Can you try updating your global nx install?

Other than that, its hard to tell why this could be happening. When we launch a task, we essentially tell VSCode what to run. We don't do anything special other than build the command. 

So with that, lets try running a command through the same mechanism Nx Console does but manually configure it. 
1. create a `tasks.json` file in your `.vscode` directory in the workspace
2. Copy and paste this:
```
{
  "version": "2.0.0",
  "tasks": [
    {
      "type": "shell",
      "problemMatcher": [],
      "command": "yarn nx run cart:lint",
      "label": "shell: nx",
      "detail": "nx"
    }
  ]
}
```
3. cmd+p > then type `task shell: nx` > execute the task
<img width="617" alt="image" src="https://github.com/nrwl/nx-console/assets/4332460/acd4a3dd-55a1-4a6a-b5f7-e4a40ffd66ee">

Does everything run fine with that?

**Comment by joeldbirch** [comment]
Thanks so much for your quick reply. Unfortunately no, when I reproduce the steps you suggested, I still get the same "Missing Platform Dependency" error. Nor did updating my global NX improve the situation.

I'll try appeasing the "or your system is missing some system dependencies" part of the error and update my OS, etc.

```
Executing task: yarn nx run cart:lint 

yarn run v1.22.19
$ nx run cart:lint

 >  NX   Missing Platform Dependency

   The Nx CLI could not find or load the native binary for your supported platform (darwin-x64).
   This likely means that optional dependencies were not installed correctly, or your system is missing some system dependencies.
   For more information please see https://nx.dev/recipes/ci/troubleshoot-nx-install-issues
   ```

**Comment by joeldbirch** [comment]
I just downloaded the previous version of VS Code (1.79.2) and the error does not occur. So it's definitely something to do with the recent VS Code update, somehow.

**Comment by joeldbirch** [comment]
I wonder if it's [something similar to this issue](https://github.com/microsoft/vscode-dotnettools/issues/141), which I think implies that the wrong version of the NX Console extension is getting installed…?

Incidentally, I remember having this issue last week with the Insiders version of VS Code, which caused me to switch back to the regular version, so that's another confirmation that something in the 1.80 update of VS Code is at play here.

**Comment by joeldbirch** [comment]
I read that VS Code try to avoid breaking existing extensions, so [I have reported this issue there](https://github.com/microsoft/vscode/issues/187332), in case it's something they need to fix.

**Comment by gersongams** [comment]
I'm also having the same issue with a m1 😢

**Comment by Cammisuli** [comment]
@joeldbirch if you're getting the same error message with the VSCode task, then it's definitely something with VSCode. 

@gersongams did you do the same steps I posted in https://github.com/nrwl/nx-console/issues/1810#issuecomment-1625325983? Does the task run there, or no?

**Comment by joeldbirch** [comment]
@Cammisuli thank you, it certainly does seem so. Have been keenly tracking the movement on the Microsoft issue I created.

@gersongams I've unblocked myself in the meantime by turning off auto-updates in VS Code's settings (search for the "update mode" setting) then installing the previous version 1.79.2.

**Comment by joeldbirch** [comment]
I’m not sure how closely they read my issue. They’ve hand balled it back to an extension issue. https://github.com/microsoft/vscode/issues/187332#issuecomment-1630348652

**Comment by MaxKless** [comment]
Yes I think it was framed a little too much extension-specific. Since you can reproduce the issue with only the Tasks API, you could create a new issue with a minimal repro that doesn't involve Nx Console. 
Then they will not just close it like that.

**Comment by joeldbirch** [comment]
Good idea, thank you for the tip! I'll do that.

**Comment by joeldbirch** [comment]
Some interesting finds while creating a stripped-down test case:

This WORKS in both the problematic VS Code 1.80.0 and previous version 1.79.2…
```jsonc
// .vscode/tasks.json
{
  "version": "2.0.0",
  "tasks": [
    {
      "type": "shell",
      "problemMatcher": [],
      "command": "echo 'basic test ran'", // <--- no NX use
      "label": "shell: debug",
      "detail": "debug"
    }
  ]
}
```

This does NOT work in VS Code 1.80.0…
```jsonc
{
  "version": "2.0.0",
  "tasks": [
    {
      "type": "shell",
      "problemMatcher": [],
      "command": "nx run cart:debug", // <-- "debug" is a package.json script with "echo 'basic test ran'" in it.
      "label": "shell: debug",
      "detail": "debug"
    }
  ]
}
```

Interestingly, this DOES work in VS Code 1.80.0…
```jsonc
{
  "version": "2.0.0",
  "tasks": [
    {
      "type": "shell",
      "problemMatcher": [],
      "command": "pnpx nx run cart:debug", // <-- using nx via pnpx
      "label": "shell: debug",
      "detail": "debug"
    }
  ]
}
```
So it seems to be a combination of VS Code version >1.80.0's task API with NX scripts, but not when pnpx triggers the NX script. I'm at a bit of a loss.

**Comment by joeldbirch** [comment]
@Cammisuli from [your comment in another thread](https://github.com/nrwl/nx/issues/17767#issuecomment-1628812403) I tried adding `--force` to `pnpm install` and everything seems to work on VS Code >= 1.80.0. I'm not thrilled with the workaround, though.

**Comment by Cammisuli** [comment]
Thats interesting. `--force` basically installs all the optional dependencies. But you already get the darwin package..

**Comment by joeldbirch** [comment]
I already got the darwin-arm package, not the darwin-x64 package, which from the error message seems to be how the process is incorrectly identifying my machine. I suppose forcing that one to download allows it to work.

**Comment by joeldbirch** [comment]
My experiments with the VS Code task API (above) left me confused as to whether it's right to post another issue to Microsoft about it. Seems to be the platform detection code (I don't know where that is; maybe a dependency of NX?) but somehow also with a change in VS Code 1.80.0 (although Microsoft claim "no changes to the runtime in 1.80.0").

**Comment by Cammisuli** [comment]
Oh I didn't realize that vscode was trying to use the x64 version instead of the arm one. In that case, there might be some application or process that's using the rosetta translation to run x64 apps. 

Can you try to find that rouge application? You can use the activity monitor to see what is running with the "Intel" cpu kind, like so:
 
<img width="1125" alt="image" src="https://github.com/nrwl/nx-console/assets/4332460/6b8a366a-a8ff-456c-b40a-a9eb8fa8a092">

**Comment by joeldbirch** [comment]
Thank you. I sorted the processes by "kind" and force-quit everything I could, restarted VS Code, then retested the task API thing. No change - I still get the "Missing Platform Dependency" error. The only Intel processes that auto restarted prior to the test were these:
![Screenshot 2023-07-14 at 9 33 05 am](https://github.com/nrwl/nx-console/assets/75602/7dfef9fb-25b5-42e4-ae8e-13f2ded17fee)

**Comment by plusnia-xebia** [comment]
Hi, I have a similar issue with Husky and nx. Inside pre-commit I want to run `nx test` and  before commit I have an error:

 The Nx CLI could not find or load the native binary for your supported platform (darwin-x64).

**Comment by sourav-bhar** [comment]
> I just downloaded the previous version of VS Code (1.79.2) and the error does not occur. So it's definitely something to do with the recent VS Code update, somehow.

Experienced the same issue after updating VS Code and downgrading to the 1.79.2 removes this issue for me as well.

**Comment by joeldbirch** [comment]
> @joeldbirch if you're getting the same error message with the VSCode task, then it's definitely something with VSCode.

@Cammisuli I'm not sure if this is still your position, but the Microsoft devs are asking for more info from you about why this it's a VS Code issue and not an extension issue. Would you be able to to [chime-in over on that thread](https://github.com/microsoft/vscode/issues/187332#issuecomment-1637663109) please?

> @joeldbirch It would be great and if you can ask the extension team why it is a VS Code bug and if they have the root cause.

**Comment by andyjessop** [comment]
I'm seeing this error on `npm install` after upgrading to `@nx/` from `@nrwl/`.

One thing to note is that `@nrwl/cli` didn't get upgraded:

My current package.json
```
    "@nrwl/cli": "15.4.4",
    "@nx/devkit": "16.5.2",
```

I'm going to try to upgrade that to `@nx/cli 16.5.2` and see if it works.

**Comment by plusnia-xebia** [comment]
it helped with my problems: https://github.com/lovell/sharp/issues/2678#issuecomment-827310464

**Comment by andyjessop** [comment]
Ok, I believe that the issue is that `"@nrwl/cli": "15.4.4",` **should have been removed from the package.json during this migration**. Can someone from the Nx team please confirm?

If I remove that from package.json, then I'm able to install correctly.

**Comment by karol1500** [comment]
Try installing `@nx/nx-linux-x64-musl` as an optional dependency. That's what solved the problem for me. Please notice that it was previously scoped to `@nrwl`. You may need to adjust that in case `@nrwl/nx-linux-x64-musl` is already installed

**Comment by joeldbirch** [comment]
@elvisun That's strange because (as far as I can recall) the task API seems to use a version of Node internal to VS Code, and any version I change to (via `nvm`) affected that.

**Comment by elvisun** [comment]
> @elvisun That's strange because (as far as I can recall) the task API seems to use a version of Node internal to VS Code, and any version I change to (via `nvm`) affected that.

Actually that's right, `nvm use` 16 or 18 worked, so can confirm it's due to VS code using an internal version of node

**Comment by joeldbirch** [comment]
> @elvisun That's strange because (as far as I can recall) the task API seems to use a version of Node internal to VS Code, and any version I change to (via nvm) affected that.

Sorry, I made a typo. I meant "any version I change to (via nvm) **doesn't** affect that."

**Comment by jmaylin** [comment]
**Upgrading node from 16 to 18 on the default mac terminal solved the issue for me**

More details

On a daily basis, I work with iTerm (node 18), which explains why I can run `yarn nx run my-project:serve`

But I guess VScode relies on the default mac terminal (node 16). That can explain why I cannot launch the same project using the vscode plugin

**Comment by Cammisuli** [comment]
Does the problem still occur if anyone uses the insiders build? Maybe a fresh install might kick it into gear

**Comment by joeldbirch** [comment]
@Cammisuli I tried that, no luck. I'm about to go into a workshop to introduce the wider dev teams to my NX monorepo. This part of it is going to be awkward.

**Comment by MaxKless** [comment]
Did you try the solution posted by @jmaylin?

**Comment by TomWhiteIMO** [comment]
I had this problem because I use NVM.

My original 'system' Node version before I installed NVM is the one that was being used by the VSCode task, which seems silly but is actually the one that appears in your PATH before NVM takes over.

I fixed this problem by `nvm use system` which switches to the base Node version and then downloading and running a Node installer for [lts/hydrogen](https://nodejs.org/en).

These links gave me clues:
https://stackoverflow.com/questions/44700432/visual-studio-code-to-use-node-version-specified-by-nvm
https://bobbyhadz.com/blog/set-default-node-version-with-nvm#seeing-the-old-nodejs-version-in-vs-code

**Comment by joeldbirch** [comment]
Solved! Thank you @TomWhiteIMO for your comment and links.

By changing the tasks.json debugging command ([suggested far above](https://github.com/nrwl/nx-console/issues/1810#issuecomment-1625325983) by @Cammisuli - thank you) to run `which node` I was able to see the path to the problematic Node installation VS Code's task API was using. In my case, it was from a `brew` installation that I think I had aliased to `/usr/local/bin/node` at some point. Also, it's likely that that version was installed on an Intel machine prior to moving to a new M2 Mac, which might explain the incorrect platform detection.

**Tl;dr:**
I ran `rm -rf /usr/local/bin/node` then the NX Console tasks immediately worked because VS Code used the preferred `nvm` installation of Node.

Here is the project's .vscode/tasks.json file I used to help debug this. You can run via Cmd-P then type `task shell`. It will output the path to the Node that VS Code uses to run the task.

```json
{
  "version": "2.0.0",
  "tasks": [
    {
      "type": "shell",
      "problemMatcher": [],
      "command": "which node",
      "label": "shell: nx",
      "detail": "nx"
    }
  ]
}
```

**Comment by shyaaam** [comment]
For me, it still showed error even when my terminal (and vscode terminal) showed node v.18.17.0

SO i realized that the problem is actually if i use pnpm in my nx repo. I removed pnpm, deleted node_modules and pnpm-lock.yml file, re-installed all packages with npm, and it worked.

**Comment by MaxKless** [comment]
could you let me know the error? pnpm is supported by nx

**Comment by shyaaam** [comment]
Its the same error as OP posted. 

```
>  NX   Missing Platform Dependency

   The Nx CLI could not find or load the native binary for your supported platform (darwin-arm64).
   This likely means that optional dependencies were not installed correctly, or your system is missing some system dependencies.
   For more information please see https://nx.dev/recipes/troubleshooting/troubleshoot-nx-install-issues
```

**Comment by shyaaam** [comment]
With npm, i find no such issues in my m1 mac mini. 
But weirdly, in my m1 macbook pro, it all works fine with pnpm. I was trying to pull the same repo that works on my mb pro, to my m1 mac mini. With same plugins version (same package.json file) and everything, pnpm doesnt work on mac mini. Only npm does.

**Comment by shyaaam** [comment]
Well, now i removed both npm and pnpm, and tried the new kid in the town -  "bun install". All working perfectly fine! And also lightning fast package installs. 

I think i will stick to Bun.

**Comment by ekkolon** [comment]
Had the same issue on WSL 2 (Debian) using NVM.

I did the following:

1. Install `@nx/nx-linux-arm64-musl` package as DEV dependency
2. Install NX Console VSCode extension

The error disappeared and I was able to run `nx` commands successfuly.

I cannot tell though whether it worked because I installed the`@nx/nx-linux-arm64-musl` package or because of the NX Console extension

---

**OS**: Linux - WSL 2 (Debian)
**IDE**: VSCode 1.82.2
**NVM**: 0.39.5
**Node**: 18.15.0
**NX**: 16.9.0
**Package Manager**: PNPM

**Comment by weilinzung** [comment]
> Solved! Thank you @TomWhiteIMO for your comment and links.
> 
> By changing the tasks.json debugging command ([suggested far above](https://github.com/nrwl/nx-console/issues/1810#issuecomment-1625325983) by @Cammisuli - thank you) to run `which node` I was able to see the path to the problematic Node installation VS Code's task API was using. In my case, it was from a `brew` installation that I think I had aliased to `/usr/local/bin/node` at some point. Also, it's likely that that version was installed on an Intel machine prior to moving to a new M2 Mac, which might explain the incorrect platform detection.
> 
> **Tl;dr:** I ran `rm -rf /usr/local/bin/node` then the NX Console tasks immediately worked because VS Code used the preferred `nvm` installation of Node.
> 
> Here is the project's .vscode/tasks.json file I used to help debug this. You can run via Cmd-P then type `task shell`. It will output the path to the Node that VS Code uses to run the task.
> 
> ```json
> {
>   "version": "2.0.0",
>   "tasks": [
>     {
>       "type": "shell",
>       "problemMatcher": [],
>       "command": "which node",
>       "label": "shell: nx",
>       "detail": "nx"
>     }
>   ]
> }
> ```


This command fixed for me, basically i am using NVM~~~
```
rm -rf /usr/local/bin/node
```

**Comment by otostogan** [comment]
SOLVED 
sudo nx serve:app

**Comment by Benjythebee** [comment]
This is happening all the time for me on Windows 11;
I have to manually delete package-lock.json and @nx in node_modules, and then run `npm i` again to then be able to run my project again

**Comment by MaxKless** [comment]
Hey @Benjythebee thanks for the report. This issue is old and convoluted so it's hard to keep track of what's going on.
Could you open a new issue with more details and a description of the problem you're facing?
I'll lock this for further discussion.
